File: assets/packs/stronghold/scripts/units/peasant.lua
-- Peasant Unit Script
-- Basic worker unit for the stronghold simulation
-- Based on stronghold_old/scripts/units/military.lua peasant definition

local peasant = {
    id = "peasant",
    name = "Peasant",
    category = "civilian",
    
    -- Cost
    gold_cost = 0,  -- Free, comes from population
    
    -- Stats
    health = 10,
    speed = 5,
    
    -- Work capabilities
    can_work = true,
    can_construct = true,
    can_repair = true,
    can_firefight = true,
    can_haul = true,
    
    -- Combat (when drafted)
    attack_damage = 2,
    defense = 0,
    morale = 2,
    
    -- Resources they can gather
    can_gather = {
        "wood",
        "stone", 
        "iron",
        "food"
    },
    
    -- Basic needs
    needs_food = true,
    needs_shelter = true
}

-- Called when peasant is spawned
function peasant:on_spawn(world, x, y)
    print("Peasant spawned at (" .. x .. ", " .. y .. ")")
    return true
end

-- Called each update tick
function peasant:on_update(world, dt)
    -- Basic survival needs
    -- This would be handled by the simulation system
    return true
end

-- Called when peasant performs work
function peasant:on_work_started(work_type, target)
    print("Peasant started work: " .. work_type)
    return true
end

function peasant:on_work_completed(work_type, target, result)
    print("Peasant completed work: " .. work_type .. " with result: " .. tostring(result))
    return true
end

-- Called when peasant is drafted for combat
function peasant:on_drafted()
    print("Peasant has been drafted!")
    -- Becomes a militia unit with basic combat stats
    self.category = "militia"
    self.morale = 1  -- Very low morale when drafted
    return true
end

-- Called when peasant returns from combat
function peasant:on_demobilized()
    print("Peasant returned from military service")
    self.category = "civilian"
    return true
end

return peasant

File: world_sim_simple/src/components/peasant.rs
use bevy::prelude::*;
use serde::{Deserialize, Serialize};

/// Peasant unit configuration loaded from Lua
/// Based on assets/packs/stronghold/scripts/units/peasant.lua
#[derive(Component, Clone, Debug, Serialize, Deserialize, Reflect)]
pub struct PeasantConfig {
    pub id: String,
    pub name: String,
    pub category: String,

    // Cost
    pub gold_cost: u32,

    // Stats
    pub health: f32,
    pub speed: f32,

    // Work capabilities
    pub can_work: bool,
    pub can_construct: bool,
    pub can_repair: bool,
    pub can_firefight: bool,
    pub can_haul: bool,

    // Combat (when drafted)
    pub attack_damage: u32,
    pub defense: u32,
    pub morale: u32,

    // Resources they can gather
    pub can_gather: Vec<String>,

    // Basic needs
    pub needs_food: bool,
    pub needs_shelter: bool,
}

impl Default for PeasantConfig {
    fn default() -> Self {
        Self {
            id: "peasant".to_string(),
            name: "Peasant".to_string(),
            category: "civilian".to_string(),

            // Cost
            gold_cost: 0, // Free, comes from population

            // Stats
            health: 10.0,
            speed: 5.0,

            // Work capabilities
            can_work: true,
            can_construct: true,
            can_repair: true,
            can_firefight: true,
            can_haul: true,

            // Combat (when drafted)
            attack_damage: 2,
            defense: 0,
            morale: 2,

            // Resources they can gather
            can_gather: vec![
                "wood".to_string(),
                "stone".to_string(),
                "iron".to_string(),
                "food".to_string(),
            ],

            // Basic needs
            needs_food: true,
            needs_shelter: true,
        }
    }
}

/// Component to mark an entity as a Peasant with specific configuration
#[derive(Component, Clone, Debug, Reflect)]
pub struct PeasantTag {
    pub config: PeasantConfig,
}

impl PeasantTag {
    pub fn new() -> Self {
        Self {
            config: PeasantConfig::default(),
        }
    }

    pub fn with_config(config: PeasantConfig) -> Self {
        Self { config }
    }
}

impl Default for PeasantTag {
    fn default() -> Self {
        Self::new()
    }
}
